pkg/apiserver/rest/dualwriter.go

```diff
@@ -217,6 +232,26 @@ func SetDualWritingMode(
 		}
 	}
 
+	if (desiredMode == Mode3) && (currentMode == Mode2) {
+		// This is where we go through the different gates to allow the instance to migrate from mode 2 to mode 3.
+
+		// gate #1: ensure the data is 100% in sync
+		syncOk, err := runDataSyncer(ctx, currentMode, legacy, storage, entity, reg, serverLockService, requestInfo)
+		if err != nil {
+			klog.Info("data syncer failed for mode:", m)
+		} else {
+			if syncOk {
+				// only set currentMode to 3 if data is in sync
+				currentMode = Mode3
+			}
```

Comment:
It might be safer and easier to follow if we return immediately on error vs always "updating" the current mode.

In the proposal below I did change the return behavior if there is a sync error (to return the error) and added a log entry if the sync is not ok.
```suggestion
			klog.Info("data syncer failed for mode:", m)
			return currentMode, err
		}
		if !syncOk {
			klog.Info("data syncer not ok for mode:", m)
			return currentMode, nil
		}
		// only set currentMode to 3 if data is in sync
		currentMode = Mode3
```

Reply:
Good point! I agree that it's easier to follow that way. Going to commit the suggestion.